SwiftMQTT/SwiftMQTT/MQTTReconnectingSession.swift

```diff
@@ -0,0 +1,151 @@
+			clientID: connectParams.clientID,
+			cleanSession: connectParams.cleanSession,
+			keepAlive: connectParams.keepAlive,
+			connectionTimeout: connectParams.timeout,
+			useSSL: connectParams.useSSL)
+    }
+}
+
+public class MQTTReconnectingSession: MQTTBroker {
+    fileprivate let connectParams: MQTTConnectParams
+    fileprivate let batchPredicate: ([MQTTMessage])->Bool
+    fileprivate var session: MQTTSession!
+    
+    open weak var delegate: MQTTReconnectingSessionDelegate?
+    
+    public init(connectParams: MQTTConnectParams, batchPredicate: @escaping ([MQTTMessage])->Bool) {
+        self.batchPredicate = batchPredicate
+        self.connectParams = connectParams
+    }
+    
+    public func connect(completion: MQTTSessionCompletionBlock? = nil) {
+        self.session = MQTTSession(connectParams: connectParams)
+        self.session.delegate = self
+        self.connectWithRetry(0, nil, completion)
+    }
+    
+    public func publish(_ data: Data, in topic: String, delivering qos: MQTTQoS, retain: Bool, completion: MQTTSessionCompletionBlock?) {
+        self.session.publish(data, in: topic, delivering: qos, retain: retain, completion: completion)
+    }
+    
+    public func subscribe(to topics: [String: MQTTQoS], completion: MQTTSessionCompletionBlock?) {
+        self.session.subscribe(to: topics, completion: completion)
+    }
+    
+    public func unSubscribe(from topics: [String], completion: MQTTSessionCompletionBlock?) {
+        self.session.unSubscribe(from: topics, completion: completion)
+    }
+    
+    public func disconnect() {
+        self.session = nil
+		self.delegate?.mqttConnected(false, for: self, error: nil)
+    }
+}
+
+extension MQTTReconnectingSession {
+    fileprivate func connectWithRetry(_ attempt: Int, _ error: Error?, _ completion: MQTTSessionCompletionBlock?) {
+        self.session.connect { [weak self] success, newError in
+            if success {
+                completion?(success, newError)
+                if let inform = self {
+					inform.delegate?.mqttConnected(true, for: inform, error: newError ?? error)
+                }
+            }
+            else if let retry = self {
+                retry.doRetry(attempt, newError ?? error, completion)
+            }
+            else {
+                completion?(false, newError ?? error)
+            }
+        }
+    }
+    
+    fileprivate func doRetry(_ attempt: Int, _ error: Error?, _ completion: MQTTSessionCompletionBlock?) {
+        if attempt < connectParams.retryCount && connectParams.retryTimeInterval > 0.0 {
+            DispatchQueue.global().asyncAfter(deadline: .now() + connectParams.retryTimeInterval) {
+                self.connectWithRetry(attempt + 1, error, completion)
+            }
+        }
+        else {
+            self.disconnect()
+            connectResuscitate(nil)
+        }
+    }
+    
+    fileprivate func connectResuscitate(_ completion: MQTTSessionCompletionBlock?) {
```

Comment:
Can we use more familiar terms like `reviveConnection` or `restoreConnection`? I had to look up what does `resuscitate` means 😅